File: app/src/main/java/com/mobile/rxjava2andretrofit2/java/custom_view/rclayout/BaseRCFrameLayout.java
package com.mobile.rxjava2andretrofit2.java.custom_view.rclayout;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;

/**
 * author    : Urasaki
 * e-mail    : 1164688204@qq.com
 * date      : 2020/11/12 15:30
 * introduce : 圆角子FrameLayout
 */
public abstract class BaseRCFrameLayout extends RCFrameLayout {

    private static final String TAG = "BaseRCFrameLayout";
    private Activity activity;
    protected Map<String, String> bodyParams;
    private Intent intent;
    private Bundle bundle;

    public BaseRCFrameLayout(@NonNull Context context) {
        super(context);

        this.activity = (Activity) context;
        bodyParams = new HashMap<>();
    }

    public BaseRCFrameLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);

        this.activity = (Activity) context;
        bodyParams = new HashMap<>();
    }

    protected abstract View initLayout();

    protected abstract void initData();

    protected abstract void initViews();

    protected void initViews(View view) {

    }

    protected void initLoadData() {

    }

    public void showLoading() {
    }

    public void hideLoading() {
    }

    protected void showToast(String message) {
        Toast.makeText(activity, message, Toast.LENGTH_LONG).show();
    }

    protected void startActivityCarryParams(Class<?> cls, Map<String, String> params) {
        intent = new Intent(activity, cls);
        bundle = new Bundle();

        if (params != null && params.size() > 0) {
            for (String key : params.keySet()) {
                if (params.get(key) != null) {//如果参数不是null，才把参数传给后台
                    bundle.putString(key, params.get(key));
                }
            }
            intent.putExtras(bundle);
        }
        activity.startActivity(intent);
    }

    private boolean mStartAnimator = false;

    /**
     * 设置推荐位到达边界时抖动
     *
     * @param focusView
     * @param direction 抖动方向，取值使用{@link View#FOCUS_RIGHT},{@link View#FOCUS_LEFT},
     *                  {@link View#FOCUS_UP},{@link View#FOCUS_DOWN}
     */
    protected void setShakeAnimator(View focusView, int direction) {
        ObjectAnimator translationAnimator;

        switch (direction) {
            case FOCUS_LEFT:
                translationAnimator = ObjectAnimator.ofFloat(focusView, "translationX", 0f, -15f, 0f, 15f, 0f, -15f, 0f, 15f, 0f);
                break;
            case FOCUS_DOWN:
                translationAnimator = ObjectAnimator.ofFloat(focusView, "translationY", 0f, 15f, 0f, -15f, 0f, 15f, 0f, -15f, 0f);
                break;
            case FOCUS_UP:
                translationAnimator = ObjectAnimator.ofFloat(focusView, "translationY", 0f, -15f, 0f, 15f, 0f, -15f, 0f, 15f, 0f);
                break;
            case FOCUS_RIGHT:
            default:
                translationAnimator = ObjectAnimator.ofFloat(focusView, "translationX", 0f, 15f, 0f, -15f, 0f, 15f, 0f, -15f, 0f);
                break;
        }
        translationAnimator.setInterpolator(new LinearInterpolator());
        translationAnimator.setDuration(300);
        translationAnimator.addListener(animatorListener);
        translationAnimator.start();
    }

    private Animator.AnimatorListener animatorListener = new Animator.AnimatorListener() {
        @Override
        public void onAnimationStart(Animator animation) {
            mStartAnimator = true;
        }

        @Override
        public void onAnimationEnd(Animator animation) {
            mStartAnimator = false;
        }

        @Override
        public void onAnimationCancel(Animator animation) {
            mStartAnimator = false;
        }

        @Override
        public void onAnimationRepeat(Animator animation) {
        }
    };

    protected boolean isShakeAnimationRun() {
        return mStartAnimator;
    }

}

File: library_common/src/main/java/com/phone/library_common/custom_view/rclayout/BaseRCFrameLayout.kt
package com.phone.library_common.custom_view.rclayout.helper

import android.animation.Animator
import android.animation.ObjectAnimator
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.animation.LinearInterpolator
import android.widget.Toast

abstract class BaseRCFrameLayout(context: Context) : RCFrameLayout(context) {

    private val TAG = "BaseRCFrameLayout"
    private var mActivity: Activity? = null
    protected var bodyParams: Map<String, String>? = null
    private var intent: Intent? = null
    private var bundle: Bundle? = null

    init {
        mActivity = context as Activity
        bodyParams = HashMap()
    }

    protected abstract fun initLayout(): View?

    protected abstract fun initData()

    protected abstract fun initViews()

    protected open fun initViews(view: View?) {}

    protected open fun initLoadData() {}

    open fun showLoading() {}

    open fun hideLoading() {}

    protected open fun showToast(message: String?) {
        Toast.makeText(mActivity, message, Toast.LENGTH_LONG).show()
    }

    protected open fun startActivityCarryParams(cls: Class<*>?, params: Map<String?, String?>?) {
        intent = Intent(mActivity, cls)
        bundle = Bundle()
        if (params != null && params.size > 0) {
            for (key in params.keys) {
                if (params[key] != null) { //如果参数不是null，才把参数传给后台
                    bundle!!.putString(key, params[key])
                }
            }
            intent!!.putExtras(bundle!!)
        }
        mActivity!!.startActivity(intent)
    }

    private var mStartAnimator = false

    /**
     * 设置推荐位到达边界时抖动
     *
     * @param focusView
     * @param direction 抖动方向，取值使用[View.FOCUS_RIGHT],[View.FOCUS_LEFT],
     * [View.FOCUS_UP],[View.FOCUS_DOWN]
     */
    protected open fun setShakeAnimator(focusView: View?, direction: Int) {
        val translationAnimator: ObjectAnimator
        translationAnimator = when (direction) {
            FOCUS_LEFT -> ObjectAnimator.ofFloat(
                focusView,
                "translationX",
                0f,
                -15f,
                0f,
                15f,
                0f,
                -15f,
                0f,
                15f,
                0f
            )
            FOCUS_DOWN -> ObjectAnimator.ofFloat(
                focusView,
                "translationY",
                0f,
                15f,
                0f,
                -15f,
                0f,
                15f,
                0f,
                -15f,
                0f
            )
            FOCUS_UP -> ObjectAnimator.ofFloat(
                focusView,
                "translationY",
                0f,
                -15f,
                0f,
                15f,
                0f,
                -15f,
                0f,
                15f,
                0f
            )
            FOCUS_RIGHT -> ObjectAnimator.ofFloat(
                focusView,
                "translationX",
                0f,
                15f,
                0f,
                -15f,
                0f,
                15f,
                0f,
                -15f,
                0f
            )
            else -> ObjectAnimator.ofFloat(
                focusView,
                "translationX",
                0f,
                15f,
                0f,
                -15f,
                0f,
                15f,
                0f,
                -15f,
                0f
            )
        }
        translationAnimator.interpolator = LinearInterpolator()
        translationAnimator.duration = 300
        translationAnimator.addListener(animatorListener)
        translationAnimator.start()
    }

    private val animatorListener: Animator.AnimatorListener = object : Animator.AnimatorListener {
        override fun onAnimationStart(animation: Animator) {
            mStartAnimator = true
        }

        override fun onAnimationEnd(animation: Animator) {
            mStartAnimator = false
        }

        override fun onAnimationCancel(animation: Animator) {
            mStartAnimator = false
        }

        override fun onAnimationRepeat(animation: Animator) {}
    }

    protected open fun isShakeAnimationRun(): Boolean {
        return mStartAnimator
    }

}